src/lib/transition_handler/catchup_scheduler.ml

```diff
@@ -84,37 +107,79 @@ module Make (Inputs : Inputs.S) = struct
+          Hashtbl.add_exn t.parent_root_timeouts ~key:parent_hash
+            ~data:(make_timeout ()) ;
+          [transition]
+      | Some sibling_transitions ->
           if
-            List.exists entries ~f:(fun (trans, _) ->
-                State_hash.equal hash (With_hash.hash trans) )
+            List.exists sibling_transitions ~f:(fun collected_transition ->
+                State_hash.equal hash @@ With_hash.hash collected_transition )
           then (
             Logger.info t.logger
               !"Received request to watch transition for catchup that already \
                 was being watched: %{sexp: State_hash.t}"
               hash ;
-            entries )
-          else (transition, make_timeout ()) :: entries )
+            sibling_transitions )
+          else (
+            cancel_child_timeout t hash ;
+            Hashtbl.add_exn t.collected_transitions ~key:hash ~data:[] ;
+            transition :: sibling_transitions ) )
 
-  let rec extract t (transition, timeout) =
-    Time.Timeout.cancel t.time_controller timeout () ;
+  let rec extract t transition =
     let successors =
       Option.value ~default:[]
-        (Hashtbl.find t.timeouts (With_hash.hash transition))
+        (Hashtbl.find t.collected_transitions (With_hash.hash transition))
     in
     Rose_tree.T (transition, List.map successors ~f:(extract t))
 
+  let rec remove_tree t parent_hash =
+    let children =
+      Option.value ~default:[]
+        (Hashtbl.find t.collected_transitions parent_hash)
+    in
+    Hashtbl.remove t.collected_transitions parent_hash ;
+    List.iter children ~f:(Fn.compose (remove_tree t) With_hash.hash)
+
+  (*
+  let notify t ~transition =
+    let hash = With_hash.hash transition in 
+    if (Option.is_none @@ Hashtbl.find t.parent_root_timeouts hash) &&
+       (Option.is_some @@ Hashtbl.find t.collected_transitions hash)
+    then Logger.info t.logger !"Received notify on node that is waiting for catchup: %{sexp: State_hash.t}" hash
+    else
+      cancel_timeout t hash ;
+      Option.iter (Hashtbl.find t.collected_transitions hash)
+        ~f:(fun collected_transitions ->
+          let transition_branches = List.map collected_transitions ~f:(extract t)
+          in Capped_supervisor.dispatch t.breadcrumb_builder_supervisor transition_branches ) ;
+      remove_tree t hash
+*)
+
   let notify t ~transition =
-    Option.iter
-      (Hashtbl.find t.timeouts (With_hash.hash transition))
-      ~f:(fun entries ->
-        Hashtbl.remove t.timeouts (With_hash.hash transition) ;
-        let transition_branches = List.map entries ~f:(extract t) in
-        Capped_supervisor.dispatch t.breadcrumb_builder_supervisor
-          transition_branches )
+    let hash = With_hash.hash transition in
+    if
+      (Option.is_none @@ Hashtbl.find t.parent_root_timeouts hash)
+      && (Option.is_some @@ Hashtbl.find t.collected_transitions hash)
+    then
+      Or_error.errorf
```

Comment:
This is a bit confusing. I would say "Received notification to kill catchup job on a non-parent_root_transition"

Reply:
Changed as requested.